WIP:  Blog Post Feature

parsa-tehranipoor opened the pull request:
### Summary <!-- Required -->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Itemize bug fixes, new features, and other changes -->
<!-- Feel free to break this into sub-sections, i.e. features, fixes, etc. -->
<!-- Some examples are shown below. -->

This PR focuses on both the frontend and the backend for my blog post feature. The feature is still a work in progress and will be completed sometime by the end of the semester.

So far, most of the frontend has been completed, with only one task left to complete. For the frontend on the admin side, which would allow for the team to actually create and post blogs, all that's left is to polish parts of the blog post creation feature and start on the blog post activation portal for archiving or posting blogs

- [x] Implemented home page for showcasing blogs
- [x] Developed design for blog post content
- [ ] Code up admin blog post creation
- [ ] Implement blog post activation portal

The backend is now complete. Every route has been created, documented, and tested, and some older routes were tweaked in conjunction with the change in our database schema

- [x] Implemented creation, deletion, and update routes
- [x] Designed database scheme
- [x] Implement additional routes
- [x] Add Documentation
- [x] Migrated collection data in 'likes' and 'users' collections
- [x] Finish Testing routes

### Test Plan <!-- Required -->

<!-- Provide screenshots or point out the additional unit tests -->
To test the routes, I used Postman for each route and checked them based on the changes made in the database. I turned of authentication for each route to simplify the process. Here is a picture of all the routes I have tested so far on Postman
<img width="1512" height="551" alt="Screenshot 2025-10-26 at 11 44 42 PM" src="https://github.com/user-attachments/assets/d8abc9fb-24c5-413c-bc2f-e76cff3c3715" />

And here are some preliminary screenshots of the current state of the frontend

<img width="1511" height="943" alt="Screenshot 2025-12-07 at 10 34 29 PM" src="https://github.com/user-attachments/assets/4901c7c7-ab84-4c4d-9e59-49f342a0ee8e" />
<img width="1512" height="949" alt="Screenshot 2025-12-07 at 10 34 38 PM" src="https://github.com/user-attachments/assets/e27fa7f9-a5d1-41cf-af80-f5dcfb84ebb6" />
<img width="1512" height="941" alt="Screenshot 2025-12-07 at 10 34 49 PM" src="https://github.com/user-attachments/assets/d5856095-712e-458e-83ee-a1b46e599537" />

parsa-tehranipoor committed:
Added more routes plus documentation

laurenp-2 commented:
This is a great step towards the finished blog feature! The api routes are really organized and easy to read—the documentation is especially great and will be helpful to reference later on. Can't wait to see the finished feature!

parsa-tehranipoor committed:
Finished testing, migrated data in firebase, and tweaked routes accordingly

CasperL1218 reviewed (changes requested):
1. For this `new-blog-post` endpoint, the current implementation is using the 401 error code also for handling missing content fields, which is not covered in the documentation. This should be handled either with a different error code or you should adjust the endpoint spec to reflect this. Here is the current [endpoint specification.](https://github.com/cornell-dti/cu-apts/pull/411/files#diff-8455472d7e00dd51813de7e63e5b214382a247990f84bf3fbed9fcc3af24ad6fR85)

2. For the `delete-blog-post` endpoint, the [comments](https://github.com/cornell-dti/cu-apts/pull/411/files#diff-8455472d7e00dd51813de7e63e5b214382a247990f84bf3fbed9fcc3af24ad6fR135) mentioned that it will be checking for either admin access or blog post creator access. However, the current code is only conducting admin check, which to be honest should be enough considering our current plan of use for this feature. Nevertheless, it would be a good idea to consider adding a creator field to specify the ownership of the blog post and conduct corresponding authentication checks.

3. For the `edit-blog-post` [endpoint](https://github.com/cornell-dti/cu-apts/pull/411/files#diff-8455472d7e00dd51813de7e63e5b214382a247990f84bf3fbed9fcc3af24ad6fR172), I believe you have commented out the authentication portions (possibly for testings). Remember to uncomment them later.

4. The `userId` field in the [type definition](https://github.com/cornell-dti/cu-apts/pull/411/files#diff-93f142ce065fbfac56abb3673fd7cbb72063360f52ed382fc2df4684ec5fb319R44) for the blog post need to be clearer. I assume it should be referring to the creator of the blog post, so you should rename it to be more specific.  


Overall, the structure of the endpoints are well-organized. The update for the handle-like endpoint for different targets are well done. Keep up with the detailed documentations and be careful about consistency between them and the actual implementation. Good job!

laurenp-2 commented on backend/src/app.ts:
```diff
@@ -68,6 +71,263 @@ app.get('/api/faqs', async (_, res) => {
+      !blogPost.tags ||
+      !blogPost.visibility
+    ) {
+      res.status(401).send('Error: missing fields');
```
Although an error is raised, since its not returned the code will continue and create a doc anyway (also, in the spec you say 401 is due to authentication issues—400 Bad Request might fit better)

laurenp-2 commented on backend/src/app.ts:
```diff
@@ -68,6 +71,263 @@ app.get('/api/faqs', async (_, res) => {
+ * - 401: Error due to unauthorized access or authentication issues.
+ */
+app.post('/api/edit-blog-post/:blogPostId', async (req, res) => {
+  // if (!req.user) {
```
if testing is complete don't forget to uncomment this!

laurenp-2 commented on backend/src/app.ts:
```diff
@@ -68,6 +71,263 @@ app.get('/api/faqs', async (_, res) => {
+  }
+  try {
+    // Update the status of the blog post document to 'DELETED'
+    await blogPostCollection.doc(blogPostId).update({ status: 'ARCHIVED' });
```
There seems to be a mismatch between archiving and deleting blog posts. It might be helpful to describe the difference in the specifications!

laurenp-2 reviewed (commented):
This is overall a great foundation for the BlogPosts! The most important changes are to make sure that your endpoints match the specifications and that all behavior is documented (so making sure that the http status codes are reflected in the specification and also that the archive vs delete behavior is explained clearly in documentation). Overall super excited to see the full feature!

CasperL1218 committed:
1. package.json (root):
    - Updated engines.node from "16.x" to ">=18"
    - Added resolutions to pin React types: @types/react: 17.0.80, @types/react-dom: 17.0.25
  2. frontend/package.json:
    - Updated react-scripts from 4.0.0 to ^5.0.1
    - Updated typescript from ~4.0.5 to ^5.0.0
    - Updated @types/react from ^16.9.53 to 17.0.80
    - Updated @types/react-dom from ^16.9.8 to 17.0.25
    - Added eslint-config-react-app
  3. backend/package.json:
    - Updated typescript from ^4.0.5 to ^5.0.0
  4. frontend/src/utils/firebase.ts:
    - Removed unnecessary dotenv import and config() call
  5. frontend/src/components/Apartment/MapInfo.tsx:
    - Removed unnecessary dotenv import and config() call
  6. frontend/src/components/Search/SearchResultsMap.tsx:
    - Removed unnecessary dotenv import and config() call
  7. frontend/src/react-app-env.d.ts:
    - Added global firebase.User type declaration for TypeScript compatibility
  8. yarn.lock:
    - Updated with new dependency versions

CasperL1218 committed:
Files changed:
  - .github/workflows/ci.yml - Node 16 → 18
  - .github/workflows/firebase-hosting-merge.yml - Added Node 18 setup step
  - .github/workflows/firebase-hosting-pull-request.yml - Node 16 → 18

CasperL1218 committed:
format changes

CasperL1218 committed:
fix lint error

parsa-tehranipoor committed:
Fixed dependency issues, edited AdminPage to include blog post creation, and continued work on blog post frontend

parsa-tehranipoor committed:
Add new developer Parsa Tehranipoor to Contributors list on README file

parsa-tehranipoor committed:
Continued to work on the admin side of the blog post feature via UI fixes and adding templates, also created blog post detail pages where content can be viewed and made minor tweaks to some routes

parsa-tehranipoor committed:
Added documentation to BlogPostPage and BlogPostDetailPage files

parsa-tehranipoor committed:
Completed BlogPost Admin page for adding new blogposts
